src/components/Balance/BalanceInput/MetricUnitSelector/hooks/useModalPortalElement.tsx

```diff
@@ -0,0 +1,42 @@
+import { useCallback, useMemo } from 'react';
+import { AssetBalanceInputProps } from '../../../AssetBalanceInput/AssetBalanceInput';
+import { ModalPortalElementFactory } from '../../../AssetBalanceInput/hooks/useModalPortal';
+import { CloseModal } from '../../../AssetBalanceInput/hooks/useModalPortalElement';
+import { MetricUnit } from '../../../metricUnit';
+import { MetricUnitItem } from '../MetricUnitItem/MetricUnitItem';
+import { MetricUnitSelectorProps } from '../MetricUnitSelector';
+
+export type ModalPortalElement =
+    ({ units, onUnitSelected, unit }: Pick<MetricUnitSelectorProps, 'units' | 'onUnitSelected' | 'unit'>)
+    => ModalPortalElementFactory
+
+export const useModalPortalElement: ModalPortalElement = ({ units, onUnitSelected, unit }) => {
+    const handleUnitSelected = useCallback((closeModal: CloseModal) => (
+        (unit: MetricUnit) => {
+            closeModal();
+            onUnitSelected(unit);
+        }
+    ), [onUnitSelected]);
+
+    const activeUnit = useMemo(() => unit, [unit]);
+
+    return useCallback(({ closeModal, elementRef, isModalOpen }) => {
+        return <div className='metric-unit-selector__unit-list flex-container column' ref={elementRef}>
+            {isModalOpen
+                ? <>
+                    {units?.map((unit, i) => (
+                        <MetricUnitItem
+                            key={i}
```

Comment:
populate `key` with a unique value from `unit`,  like in my comment on that other `map` 👍